Tests/LifecycleTests/ComponentLifecycleTests.swift

```diff
@@ -839,4 +840,148 @@ final class ComponentLifecycleTests: XCTestCase {
+
+        for index in 0 ..< count {
+            let item = Item(label: "\(index)", failStart: index == count / 2, exptectedState: index <= count / 2 ? .shutdown : .idle, sempahore: sempahore)
+            lifecycle.register(label: item.label, start: .sync(item.start), shutdown: .sync(item.shutdown))
+        }
+
+        lifecycle.start { error in
+            XCTAssertNotNil(error, "expecting error")
+            lifecycle.shutdown()
+        }
+        lifecycle.wait()
+
+        XCTAssertEqual(.success, sempahore.wait(timeout: .now() + 1))
+    }
+
+    func testShutdownIfNotStartedWhenAsked() {
+        class DestructionSensitive {
+            let label: String
+            let failStart: Bool
+            let sempahore: DispatchSemaphore
+            var state = State.idle
+
+            deinit {
+                XCTAssertEqual(self.state, .shutdown, "\"\(self.label)\" should be shutdown")
+                self.sempahore.signal()
+            }
+
+            init(label: String, failStart: Bool = false, sempahore: DispatchSemaphore) {
+                self.label = label
+                self.failStart = failStart
+                self.sempahore = sempahore
+            }
+
+            func start() throws {
+                self.state = .started
+                if self.failStart {
+                    self.state = .error
+                    throw InitError()
+                }
+            }
+
+            func shutdown() throws {
+                self.state = .shutdown
+            }
+
+            enum State {
+                case idle
+                case started
+                case shutdown
+                case error
+            }
+
+            struct InitError: Error {}
+        }
+
+        let sempahore = DispatchSemaphore(value: 6)
+        let lifecycle = ServiceLifecycle(configuration: .init(shutdownSignal: nil))
+
+        let item1 = DestructionSensitive(label: "1", sempahore: sempahore)
+        lifecycle.register(label: item1.label, start: .sync(item1.start), shutdown: .sync(item1.shutdown))
+
+        let item2 = DestructionSensitive(label: "2", sempahore: sempahore)
+        lifecycle.registerShutdown(label: item2.label, .sync(item2.shutdown))
+
+        let item3 = DestructionSensitive(label: "3", failStart: true, sempahore: sempahore)
+        lifecycle.register(label: item3.label, start: .sync(item3.start), shutdown: .sync(item3.shutdown))
+
+        let item4 = DestructionSensitive(label: "4", sempahore: sempahore)
+        lifecycle.registerShutdown(label: item4.label, .sync(item4.shutdown))
+
+        let item5 = DestructionSensitive(label: "5", sempahore: sempahore)
+        lifecycle.register(label: item5.label, start: .none, shutdown: .sync(item5.shutdown))
+
+        let item6 = DestructionSensitive(label: "6", sempahore: sempahore)
+        lifecycle.register(label: item6.label, shutdownIfNotStarted: true, start: .sync(item3.start), shutdown: .sync(item6.shutdown))
```

Comment:
Mixed feelings here tbh.

I would love to rather revisit the design I proposed a while ago which allows the shutdown to get values from the start. This way, if there was no start called there is no start value.

I.e. 

```swift
// start, semantically equivalent to:  
//     () -> SomeValue / EventLoopFuture<SomeValue>
// shutdown, semantically equivalent to: 
//    (TaskStartResult<SomeValue, Error>) -> Void / EventLoopFuture<Void>
lifecycle.register(
    label: item6.label, 
    start: .sync(DestructionSensitive()), 
    shutdown: .sync { $0.success?.shutdown() }
)
```

etc... Something worth exploring? Should I prototype it or would you have a look @tomerd ?
I remain very worried about those "allocate object outside" and then mutate start/stop the thing from lifecycle callbacks without a communication channel between then somehow... (esp if threading was involved which we might get to because: https://github.com/swift-server/swift-service-lifecycle/issues/21 )

Reply:
absolutely - you are more than welcome to explore and prototype and alternative better design. note that in this case the `register` function will also need to return the instance that is created in `start` since it is needed "outside" as well.

Reply:
Ah yeah I see could be tricky... don't treat this comment as blocker -- I'll try to poke around if I could make it happen